### ISSUE TITLE
[another] an ssh installation couldn't be found  --- more patience please

### ISSUE THREAD

**Issue by HarmonyAngel** [microsoft/vscode-remote-release#3731]
No existing
 Doing

- VSCode Version: VSCodeUserSetup-x64-1.49.2.exe
- Local OS Version: win7
- Remote OS Version: linux 3.10.0
- Remote Extension/Connection Type: SSH/Docker/WSL ssh

Steps to Reproduce:
N/A
My Problem is:
    I have used the vscode remote ssh to linux development on win7 more than one year.
    But, today I find that, the "remote ssh connect to host" can not work more time, with an error as "an ssh installation couldn't be found". But only sometimes it can connect successfully without any change.
    I let ssh log level to be trace, and find that, the problem is caused by the detect to see whether ssh has been installed. It will execute "ssh -V" or "C:\windows\System32\OpenSSH\ssh.exe -V" etc. to check whether ssh has been installed. But more times the cmd cannot exit quickly which leads to failure check, and generate the error report.
    all the existing method I have tried but failed. and I have found that, the main reason is: the exec of ssh -V is so different between windows cmd shell and Git Bash shell. On Git Bash shell, the ssh -V execs so quickly and moonly, but on windows7 shell, there will be one timeout within at least 10 times, which will take at least 20 seconds to finish.
    so, How can I fixed it or let it work well with vscode remote ssh to linux platform?
Let "terminal.integrated.shell.windows": "D:\\tools\\Git\\bin\\bash.exe" seems no work.
    I have reinstalled both vscode and extensions, But still exists.
Please Help me, Thank you.
    
when remote connect successfully
21:33:39.163] Install and start server if needed
[21:33:39.206] Checking ssh with "ssh -V"
[21:33:39.552] > OpenSSH_8.3p1, OpenSSL 1.1.1g  21 Apr 2020

when failed
[21:32:49.951] Install and start server if needed
[21:32:49.962] Checking ssh with "ssh -V"
[21:32:55.043] ssh is not exiting, continuing
[21:32:55.044] Checking ssh with "C:\windows\System32\OpenSSH\ssh.exe -V"
[21:32:55.047] Got error from ssh: spawn C:\windows\System32\OpenSSH\ssh.exe ENOENT

Does this issue occur when you try this locally?: Yes/No  N/A
Does this issue occur when you try this locally and all extensions are disabled?: Yes/No N/A

**Comment by roblourens** [comment]
It should not take that long to run `ssh -V`, is there something going on in the background that slows it down? You can see that when it works, it only takes 300ms

**Comment by Sigrar** [comment]
I also can reproduce this problem. Sometimes it will take several seconds to get the result of the cmd `ssh -V`. But with Git bash it will show the version of ssh immediately.

**Comment by HarmonyAngel** [comment]
> It should not take that long to run `ssh -V`, is there something going on in the background that slows it down? You can see that when it works, it only takes 300ms
It's really no workload running, And It can be reproduced very easyly after system restart.
Thank Sigrar has reproduced it.

**Comment by HarmonyAngel** [comment]
> It should not take that long to run `ssh -V`, is there something going on in the background that slows it down? You can see that when it works, it only takes 300ms

There are no workload running, and it can be reproduced very easy even after systerm restart.
Experiment on win7 cmd shell show that, ssh -V timeout will appear once every 10 times. But when use vscode remote ssh to connect linux, it will success only once every more than 10 tries. 
Need your Help Thank you.

**Comment by HarmonyAngel** [comment]
> I also can reproduce this problem. Sometimes it will take several seconds to get the result of the cmd `ssh -V`. But with Git bash it will show the version of ssh immediately.

Thank you for this experiment, and with vscode remote ssh, there will be very difficult to connect successfully even once.

**Comment by HarmonyAngel** [comment]
> I also can reproduce this problem. Sometimes it will take several seconds to get the result of the cmd `ssh -V`. But with Git bash it will show the version of ssh immediately.

Another point, this unmoonly ssh -V leads to almost unusable of the remote ssh to linux, with almost 100%  failure of ssh -V not exiting in Remote-ssh log.
Any video or screen-shotcut can be provided if you need.

**Comment by Sigrar** [comment]
Maybe this problem is related to the computer hardware configuration.  VS code SSH remote can give more time to the command `ssh -V` to get the result log.
But I have a method to avoid the error message "_an ssh installation couldn't be found_". Firstly, do the `ssh -V` in your dos cmd line. Make sure you can get the version of ssh immediately after you do the cmd ` ssh -V`. And then connect the remote host in your VS code. The interval between these two execution steps should be short. You can hava a try.

**Comment by HarmonyAngel** [comment]
> Maybe this problem is related to the computer hardware configuration. VS code SSH remote can give more time to the command `ssh -V` to get the result log.
> But I have a method to avoid the error message "_an ssh installation couldn't be found_". Firstly, do the `ssh -V` in your dos cmd line. Make sure you can get the version of ssh immediately after you do the cmd ` ssh -V`. And then connect the remote host in your VS code. The interval between these two execution steps should be short. You can hava a try.

Sorry, But it not Works. And I have worked more than one year on the same laptop and the same remote linux with the same system version, this problem happened just less than 2 weeks. IF I can let the ssh -V exec within git-bash then it will work, but I haven't found that interface.

**Comment by HarmonyAngel** [comment]
> Maybe this problem is related to the computer hardware configuration. VS code SSH remote can give more time to the command `ssh -V` to get the result log.
> But I have a method to avoid the error message "_an ssh installation couldn't be found_". Firstly, do the `ssh -V` in your dos cmd line. Make sure you can get the version of ssh immediately after you do the cmd ` ssh -V`. And then connect the remote host in your VS code. The interval between these two execution steps should be short. You can hava a try.

Correct, it works Thank you very much. But it seems that the ssh -V timeout(more than 20s) problem will happen each 1 minute, really don't know why.

**Comment by Sigrar** [comment]
> > Maybe this problem is related to the computer hardware configuration. VS code SSH remote can give more time to the command `ssh -V` to get the result log.
> > But I have a method to avoid the error message "_an ssh installation couldn't be found_". Firstly, do the `ssh -V` in your dos cmd line. Make sure you can get the version of ssh immediately after you do the cmd ` ssh -V`. And then connect the remote host in your VS code. The interval between these two execution steps should be short. You can hava a try.
> 
> Correct, it works Thank you very much. But it seems that the ssh -V timeout(more than 20s) problem will happen each 1 minute, really don't know why.

Conratulations! So maybe the root cause of this problem is the program **ssh**. Why it takes so long to get the result of `ssh -V`. Which step blocks it. Or it's just because of our envirment.

**Comment by HarmonyAngel** [comment]
> > > Maybe this problem is related to the computer hardware configuration. VS code SSH remote can give more time to the command `ssh -V` to get the result log.
> > > But I have a method to avoid the error message "_an ssh installation couldn't be found_". Firstly, do the `ssh -V` in your dos cmd line. Make sure you can get the version of ssh immediately after you do the cmd ` ssh -V`. And then connect the remote host in your VS code. The interval between these two execution steps should be short. You can hava a try.
> > 
> > 
> > Correct, it works Thank you very much. But it seems that the ssh -V timeout(more than 20s) problem will happen each 1 minute, really don't know why.
> 
> Congratulations! So maybe the root cause of this problem is the program **ssh**. Why it takes so long to get the result of `ssh -V`. Which step blocks it. Or it's just because of our environment.

Maybe the disk problems, and I will change to win10 to see whether it happens or not.

**Comment by HarmonyAngel** [comment]
> > > Maybe this problem is related to the computer hardware configuration. VS code SSH remote can give more time to the command `ssh -V` to get the result log.
> > > But I have a method to avoid the error message "_an ssh installation couldn't be found_". Firstly, do the `ssh -V` in your dos cmd line. Make sure you can get the version of ssh immediately after you do the cmd ` ssh -V`. And then connect the remote host in your VS code. The interval between these two execution steps should be short. You can hava a try.
> > 
> > 
> > Correct, it works Thank you very much. But it seems that the ssh -V timeout(more than 20s) problem will happen each 1 minute, really don't know why.
> 
> Congratulations! So maybe the root cause of this problem is the program **ssh**. Why it takes so long to get the result of `ssh -V`. Which step blocks it. Or it's just because of our environment.

I have remove almost 50GB disk But it still exists. I am trying to change to Win10 to see.

**Comment by dashesy** [comment]
ssh -V is slow if DNS is slow

**Comment by Sigrar** [comment]
> ssh -V is slow if DNS is slow

@dashesy Much appreciated for your kindly help.  I changed my DSN addr. It works~.
@HarmonyAngel  I use a proxy network. And the DSN I set to 8.8.8.8. Maybe this is the root case. Maybe you are the same issue

**Comment by bamurtaugh** [comment]
Hey @HarmonyAngel, are you still experiencing this issue? Does the comment above help your use case?

**Comment by roblourens** [comment]
> ssh -V is slow if DNS is slow

Whoa this is interesting and explains some odd past issues, do you know why this is @dashesy and how I can avoid it?

**Comment by EmbeddedBacon** [comment]
We see this constantly at my site.  I measured the time about 10 times which the time varied from 500ms to over 1000ms.  Why can't this timeout be configurable? 

`PS C:\Users_user_> Measure-Command {ssh -V}
OpenSSH_8.2p1, OpenSSL 1.1.1d 10 Sep 2019

Days : 0
Hours : 0
Minutes : 0
Seconds : 1
Milliseconds : 23
Ticks : 10238692
TotalDays : 1.1850337962963E-05
TotalHours : 0.000284408111111111
TotalMinutes : 0.0170644866666667
TotalSeconds : 1.0238692
TotalMilliseconds : 1023.8692`

**Comment by EmbeddedBacon** [comment]
There are other times we see that ssh is found, but later it can't be found

[16:42:24.031] Log Level: 2
[16:42:24.051] remote-ssh@0.55.0
[16:42:24.052] win32 x64
[16:42:24.054] SSH Resolver called for "ssh-remote+remote_machine", attempt 1
[16:42:24.054] SSH Resolver called for host:remote_machine
[16:42:24.054] Setting up SSH remote "remote_machine"
[16:42:24.101] Using commit id "2af051012b66169dde0c4dfae3f5ef48f787ff69" and quality "stable" for server
[16:42:24.103] Install and start server if needed
[16:42:24.112] Checking ssh with "ssh -V"
[16:42:24.437] > OpenSSH_for_Windows_7.7p1, LibreSSL 2.6.5

[16:42:24.471] Running script with connection command: ssh -T -D 64749 remote_machine bash
[16:42:24.479] Terminal shell path: C:\WINDOWS\System32\cmd.exe
[16:42:24.756] >

    �]0;C:\WINDOWS\System32\cmd.exe�
    [16:42:24.757] Got some output, clearing connection timeout
    [16:42:24.773] > 'ssh' is not recognized as an internal or external command,
    operable program or batch file.

[16:42:24.785] > 'ssh' is not recognized as an internal or external command,

    operable program or batch file.

It is like it found OpenSSH, but then can't execute it.

**Comment by roblourens** [comment]
If it's taking 1 second, then this is not the same issue. It already waits longer than that.

If ssh is not on the PATH in your terminal environment, then it needs to be or you need to set up the path to the executable for us to use.

**Comment by EmbeddedBacon** [comment]
Mr. @roblourens,

> If it's taking 1 second, then this is not the same issue. It already waits longer than that.

If the timeout isn't the issue, then what would cause the following output on the console if not a timeout issue?  Hung process?

> [15:24:38.893] Checking ssh with "ssh -V"
> [15:24:44.267] ssh is not exiting, continuing

>If ssh is not on the PATH in your terminal environment, then it needs to be or you need to set up the path to the executable for us to use.

If that is the case, what might cause the following log entries

[16:42:24.112] Checking ssh with "ssh -V"
[16:42:24.437] > OpenSSH_for_Windows_7.7p1, LibreSSL 2.6.5

Then get the following log entries?

C:\WINDOWS\System32\cmd.exe
[16:42:24.757] Got some output, clearing connection timeout
[16:42:24.773] > 'ssh' is not recognized as an internal or external command, operable program or batch file.

Is VS Code using two different methods of launch a task/process such that they get different responses.  We did verify from a console/cmd window that if we type "ssh -V" it does launch and output the result of launching the process.

I hope that your team can add a parameter to configure the ssh detection timeout.  Also, what other task/test/logs can the community perform to help out with these issues?  Is there a log level we can change to get more data for the team?  These issues seem to have become a painful annoyance in the past three months.  Thanks.

Mark

**Comment by roblourens** [comment]
The logs you have posted show openssh 8.2 in one place and openssh 7.7 in another place, so I assume there is something strange going on with these different versions.

Please open a new issue for the second issue, it doesn't belong here. But yes the first command is spawned from the extension host process, the second runs in a terminal, I don't think there should be a difference in environments, off the top of my head, maybe if you have a shell profile script it runs in the second but not the first.

**Comment by EmbeddedBacon** [comment]
Mr. @roblourens, yes and sorry, one log is from my machine and another log is from a co-workers machine, so that is the difference between the Open SSH versions being reported.  I will open a new one regarding ssh being found then can't be executed.

**Comment by justinbabu** [comment]
using  ssh extension  version 0.49.0 works, it seems to wait(hang) for ssh -V indefinitely.

**Comment by roblourens** [comment]
I'm just increasing the timeout here. Not really sure what else to do.

**Comment by paulhodson** [comment]
I've been having this issue too, intermittently having problems reconnecting one or more of my remote SSH windows. As noted earlier, running `ssh -V` before opening VSCode would sometimes help.

I just want to share something that seems to have worked for me. Running `strace ssh -V` and searching for some of the output from around the time of the long waits led me to some [similar ](https://github.com/git-for-windows/git/issues/193) [issues](https://stackoverflow.com/a/28427299/854089). [Installing Cygwin ](https://superuser.com/a/904600) (default install + cygrunsrv) and ensuring it auto-runs seems to have resolved it.

**Comment by sbatten** [comment]
verified by looking at change (a simple timeout bump). actual repro requires slow DNS, time will tell if this alleviates user issues in that case.
